Доработана функциональность по техническому заданию 8 спринта

asdbl opened the pull request.

asdbl committed:
Добавить время начала и длительность для задач.

asdbl committed:
Исправить ошибку CheckStyle

asdbl committed:
Исправить ошибку CheckStyle

asdbl committed:
Исправить ошибку выполнения тестов

avfyodorov commented on src/task/Epic.java:
```diff
@@ -1,15 +1,24 @@
 public class Epic extends Task {
     private final List<Integer> subtaskIdList;
+    private Duration duration;
+    private LocalDateTime startTime;
```
private Duration duration;
    private LocalDateTime startTime;

Эти два Поля не нужны, они уже есть в классе-предке Task. 
Можно использовать именно те поля. 
А так получается ненужное дублирование, которая не соответствует ООП

avfyodorov commented on src/manager/taskManager/InMemoryTaskManager.java:
```diff
@@ -236,5 +269,9 @@ public List<Subtask> getSubtasksByEpic(Epic epic) {
-}
 
+    @Override
+    public TreeSet<Task> getPrioritizedTasks() {
```
TreeSet
По Техническому заданию возвращать нужно список, а не сет: 

	@Override
	public List<Task> getPrioritizedTasks() {
		return new ArrayList<>(prioritizedTasks);
	}

avfyodorov commented on src/manager/taskManager/InMemoryTaskManager.java:
```diff
@@ -236,5 +269,9 @@ public List<Subtask> getSubtasksByEpic(Epic epic) {
 
+    @Override
+    public TreeSet<Task> getPrioritizedTasks() {
+        return new TreeSet<>(getAll());
```
Опять же по техническому заданию: 
.================================
Предполагается, что пользователь будет часто запрашивать этот список, поэтому подберите подходящую структуру данных для хранения. Сложность получения должна быть уменьшена с 
𝑂(𝑛∗𝑙𝑜𝑔(𝑛)O(n∗log(n)) до 𝑂(𝑛)O(n).
.==================================

Поэтому придётся завести отдельную структуру для хранения задач и подзадач, отсортированых по времени и не пересекающихся. Приблизительно так: 

	protected final Set<Task> prioritizedTasks = new TreeSet<>(Comparator.comparing(Task::getStartTime));

И соответственно, методы для работы с этим списком-проверить на пересечение, добавить в, удалить из.

avfyodorov commented on src/manager/taskManager/FileBackedTaskManager.java:
```diff
@@ -82,15 +85,15 @@ public String toString(Task task) {
         if (task.getClass() == Task.class) {
             tasksType = TasksType.TASK;
-            taskToString = String.format("%d,%s,%s,%s,%s", task.getId(), tasksType, task.getTaskName(), task.getStatus(), task.getTaskDescription());
+            taskToString = String.format("%d,%s,%s,%s,%s,%s,%s,%s,%s", task.getId(), tasksType, task.getTaskName(), task.getStatus(), task.getTaskDescription(), "-", task.getStartTime(), task.getDuration().toMinutes(), task.getEndTime());
```
Здесь и в аналогичных местах проекта. 
Просьба форматировать код, чтобы не было слишком длинных строк.

avfyodorov commented on src/manager/taskManager/InMemoryTaskManager.java:
```diff
@@ -29,9 +27,15 @@ public InMemoryTaskManager() {
+    }
 
+    private void timeSetter(Task task) {
+        task.setStartTime(getPrioritizedTasks().last().getEndTime().plusMinutes(15));
```
Не совсем так. Задачи и подзадачи приходят с уже установленным времени начала и продолжительностью. Эти значения ни в коем случае нельзя менять. 
А нужно проверить - нет ли пересечений с уже существующими в отсортированном списке задачами.

asdbl replied:
Не совсем понял, что тогда должно происходить при пересечении? Должна выбрасываться ошибка?

avfyodorov replied:
Да, должно выбрасываться исключение.

avfyodorov commented on src/manager/taskManager/InMemoryTaskManager.java:
```diff
@@ -29,9 +27,15 @@ public InMemoryTaskManager() {
+    private void timeSetter(Task task) {
+        task.setStartTime(getPrioritizedTasks().last().getEndTime().plusMinutes(15));
     }
 
```
Ещё один момент. 
Для эпиков. 
Нигде не нашёл расчёта времени начала, времени окончания и продолжительности Эпика . По техническому заданию: 
.--------------
 вот с классом Epic нужно поработать дополнительно. Продолжительность эпика — сумма продолжительностей всех его подзадач. Время начала — дата старта самой ранней подзадачи, а время завершения — время окончания самой поздней из задач. Новые поля duration и startTime этого класса будут расчётные — аналогично полю «статус». Для реализации getEndTime() удобно добавить поле endTime в Epic и рассчитать его вместе с другими полями.

asdbl replied:
Обновление времени эпика происходит во время добавления подзадачи в InMemoryTaskManager.
Добавил новые поля в Epic.

avfyodorov reviewed (commented):
Добрый вечер, андрей! 

В целом хорошо, правда, есть ряд уточнений. 

Cтавьте, пожалуйста, пометки, чтобы было понятно, что комментарии замечены. )

asdbl committed:
Исправить метод getPrioritizedTasks в классе InMemoryTaskManager, теперь он возвращает список.
Добавить TreeSet для хранения отсортированных задач.
Удалить метод timeSetter в классе InMemoryTaskManager.

asdbl committed:
Исправить ошибку CheckStyle.

avfyodorov commented on src/manager/taskManager/InMemoryTaskManager.java:
```diff
@@ -48,23 +51,56 @@ public void add(Epic epic) {
+        LocalDateTime endTime = subtasks.get(subLastId).getEndTime();
+        epic.setStartTime(startTime);
+        epic.setEndTime(endTime);
+        epic.setDuration(Duration.between(startTime, endTime));
```
Не совсем так. По техническому заданию: 
.=========
А вот с классом Epic нужно поработать дополнительно. Продолжительность эпика — сумма продолжительностей всех его подзадач. Время начала — дата старта самой ранней подзадачи, а время завершения — время окончания самой поздней из задач.
.==========
Посмотрите, пожалуйста, вариант решения: 
	private void updateEpicDuration(Epic epic) {
		List<Integer> subs = epic.getSubtaskIds();
		if (subs.isEmpty()) {
			epic.setDuration(0L);
			return;
		}
		LocalDateTime start = LocalDateTime.MAX;
		LocalDateTime end = LocalDateTime.MIN;
		long duration = 0L;
		for (int id : subs) {
			final Subtask subtask = subtasks.get(id);
			final LocalDateTime startTime = subtask.getStartTime();
			final LocalDateTime endTime = subtask.getEndTime();
			if (startTime.isBefore(start)) {
				start = startTime;
			}
			if (endTime.isAfter(end)) {
				end = endTime;
			}
			duration += subtask.getDuration();
		}
		epic.setDuration(duration);
		epic.setStartTime(start);
		epic.setEndTime(end);
	}

avfyodorov commented on src/manager/taskManager/InMemoryTaskManager.java:
```diff
@@ -48,23 +51,56 @@ public void add(Epic epic) {
+        }
         final Epic epicToAdd = new Epic(epic);
         epics.put(epicToAdd.getId(), epicToAdd);
+        prioritizedTasks.add(epic);
```
prioritizedTasks.add(epic);

Эпик в Список отсортированных задач добавлять не нужно. Там находятся только задачи и подзадачи. 
Например, у Эпика две подзадачи-одна подзадача сегодня и сейчас, а вторая через неделю. Между ними могут находиться другие задачи, а также подзадачи от других эпиков.

avfyodorov commented on src/manager/taskManager/InMemoryTaskManager.java:
```diff
@@ -48,23 +51,56 @@ public void add(Epic epic) {
+    public boolean isOverlap(Task task) {
+        return prioritizedTasks.stream()
+                .filter(t -> task.getStartTime().isBefore(t.getEndTime().plusMinutes(15)))
+                .anyMatch(t -> task.getEndTime().plusMinutes(15).isAfter(t.getStartTime()));
```
15
Что-то никак не могу сообразить Почему 15? 
Продолжительность задачи, а также подзадачи, в принципе, может быть любой, например 1 минута, или 3 часа.

asdbl replied:
Добавлял 15мин, чтобы между задачами был интервал, сейчас убрал.

avfyodorov reviewed (commented):
Добрый вечер, Андрей! 
Почти всё сделано, ещё чуть-чуть нужно доработать отсортированный список.

asdbl committed:
Исправить обновление времени для эпика в классе InMemoryTaskManager.
В отсортированный список теперь не добавляется эпик.

avfyodorov reviewed (commented):
Добрый день, Андрей!

Замечаний нет.
Работа принята.
